fastly/resource_fastly_service_acl_entries_v1_test.go

```diff
@@ -0,0 +1,443 @@
+			Service: service.ID,
+			Version: service.ActiveVersion.Number,
+			Name:    aclName,
+		})
+
+		if err != nil {
+			return fmt.Errorf("[ERR] Error looking up ACL records for (%s), version (%v): %s", service.Name, service.ActiveVersion.Number, err)
+		}
+
+		aclEntries, err := conn.ListACLEntries(&gofastly.ListACLEntriesInput{
+			Service: service.ID,
+			ACL:     acl.ID,
+		})
+
+		if err != nil {
+			return fmt.Errorf("[ERR] Error looking up ACL entry records for (%s), ACL (%s): %s", service.Name, acl.ID, err)
+		}
+
+		flatAclEntries := flattenAclEntries(aclEntries)
+		// Clear out the id values to allow a deep equal of the other attributes set in terraform.
+		for _, val := range flatAclEntries {
+			val["id"] = ""
+		}
+
+		sort.Slice(flatAclEntries, func(i, j int) bool {
+			return flatAclEntries[i]["ip"].(string) < flatAclEntries[j]["ip"].(string)
+		})
+
+		sort.Slice(expectedEntries, func(i, j int) bool {
+			return expectedEntries[i]["ip"].(string) < expectedEntries[j]["ip"].(string)
+		})
+
+		if !reflect.DeepEqual(flatAclEntries, expectedEntries) {
+			return fmt.Errorf("[ERR] Error matching:\nexpected: %#v\ngot: %#v", expectedEntries, flatAclEntries)
+		}
+
+		return nil
+	}
+}
+
+func testAccServiceDictionaryItemsV1Config_one_acl_no_entries(serviceName, aclName string) string {
+
+	backendName := fmt.Sprintf("%s.aws.amazon.com", acctest.RandString(3))
+	domainName := fmt.Sprintf("fastly-test.tf-%s.com", acctest.RandString(10))
+
+	return fmt.Sprintf(`
+resource "fastly_service_v1" "foo" {
+  name = "%s"
+  domain {
+    name    = "%s"
+    comment = "tf-testing-domain"
+	}
+  backend {
+    address = "%s"
+    name    = "tf -test backend"
+  }
+  acl {
+	name       = "%s"
+  }
+  force_destroy = true
+}`, serviceName, domainName, backendName, aclName)
+}
+
+func testAccServiceACLEntriesV1Config_one_acl_with_entries(serviceName, aclName string, aclEntriesList []map[string]interface{}) string {
+	backendName := fmt.Sprintf("%s.aws.amazon.com", acctest.RandString(3))
+	domainName := fmt.Sprintf("fastly-test.tf-%s.com", acctest.RandString(10))
+
+	aclEntries := ""
+
+	for _, val := range aclEntriesList {
+		aclEntries += "entry {\n"
+		aclEntries += fmt.Sprintf("ip = \"%s\"\n", val["ip"])
+		aclEntries += fmt.Sprintf("subnet = \"%s\"\n", val["subnet"])
+		aclEntries += fmt.Sprintf("negated = %t\n", val["negated"])
+		aclEntries += fmt.Sprintf("comment = \"%s\"\n", val["comment"])
+		aclEntries += "}\n"
+	}
+
+	return fmt.Sprintf(`
+variable "myacl_name" {
+	type = string
+	default = "%s"
+}
+
+resource "fastly_service_v1" "foo" {
+	name = "%s"
+	domain {
+		name    = "%s"
+		comment = "tf-testing-domain"
+	}
+	backend {
+		address = "%s"
+		name    = "tf-testing-backend"
+	}
+	acl {
+		name       = var.myacl_name
+	}
+	force_destroy = true
+}
+ resource "fastly_service_acl_entries_v1" "entries" {
+	service_id = fastly_service_v1.foo.id
+	acl_id = {for s in fastly_service_v1.foo.acl : s.name => s.acl_id}[var.myacl_name]
+	%s
```

Comment:
It might be easier to understand the configuration by applying "dynamic nested blocks" a new feature of `v0.12`. Especially because the `subnet` and the `negated` fields are static in this test. Here is an example:

```
locals {
  acl_name = "my_acl"
  acl_entries = [
    {
      ip      = "4000:db8:ffff:ffff:ffff:ffff:ffff:ffff"
      comment = "acl_entry_1"
    }
  ]
}

resource "fastly_service_acl_entries_v1" "entries" {
  service_id = fastly_service_v1.myservice.id
  acl_id     = { for d in fastly_service_v1.myservice.acl : d.name => d.acl_id }[local.acl_name]
  dynamic "entry" {
    for_each = [for e in local.acl_entries : {
      ip      = e.ip
      comment = e.comment
    }]

    content {
      ip      = entry.value.ip
      subnet  = 22
      comment = entry.value.comment
      negated = false
    }
  }
}
```

I am fine with the test definition but it took me a while to wrap my head around it. I used the "dynamic nested block" syntax for my test configuration for this PR. It might be useful to apply this syntax so that users who look at the tests find a skinnier solution if they have an ACL with multiple entries and static fields.

Reply:
So we are going to leave the test configuration function as it is and it implemented for developers.  The aim is to construct the most basic hcl code and not introduce different hcl 2 features.  The example you show has been added to the documentation.  As of writing this the docs contain and basic example followed by a more hcl2 example and the the ignore_changes example.

It feels like this is the best location for examples.